### PR TITLE
Improve BatchNorm1d training performance (CPU)

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57034 Improve BatchNorm1d performance (CUDA)
* **#57033 Improve BatchNorm1d training performance (CPU)**

CPU part of gh-38915

BatchNorm1d is implemented by looping over the channels, selecting one channel
at a time and performing cpu_serial_kernel loops per-channel. For (N, C)
contiguous layout this results in a sub-optimal strided memory access pattern;
guarunteeing no elements will ever be in the same cache line.

I fix this by passing the entire input into one `TensorIterator` and letting
it decide which dimensions to iterate over and how to divide work among threads.

For statistic updates and the backward function, I use `at::mean` and `at::sum`
instead of the ad-hoc reductions there. Not only does this allow better memory
access patterns, it also enables vectorization and so performance improves for
BatchNorm2d as well. Unfortunately, `at::var` and `at::var_mean` don't perform
as well so I've left the other reductions as they were.

Overall, on my machine this takes the 1d example from 24 ms down to 4 ms and
the 2d example from 2.5 ms down to 2 ms.

Differential Revision: [D28142333](https://our.internmc.facebook.com/intern/diff/D28142333)

### CODE REVIEW

**Comment by mruberry** [comment]
This is extremely cool, @peterbell10! I'll let @ngimel review

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -176,39 +176,42 @@ std::tuple<Tensor,Tensor,Tensor> batch_norm_cpu_transform_input_template(
...
-      cpu_serial_kernel(iter, [=](const scalar_t i) -> scalar_t {
-        return ((i - mean) * invstd) * w + b;
-      });
+  const int64_t ndim = input.dim();
+  // Helper to convert 1d tensors to an nd tensor that broadcasts with input
+  // All elements go into the channel dimension
+  DimVector sizes(ndim, 1), strides(ndim, 0);
+  auto as_nd = [&](const Tensor& t) {
+    TORCH_INTERNAL_ASSERT(t.defined() && t.dim() == 1);
+    sizes[1] = t.sizes()[0];
```
Comment (ngimel) [inline aten/src/ATen/native/Normalization.cpp:185]: strictly speaking you need to be sure that ndim is >0 here, but I trust there was some check to that effect before batch_norm_cpu_transform_input_template was called

**Review by ngimel — APPROVED** [review]
This is a good simplification and perf improvement! One day, `var` in core will also be fast so you won't have to have a special kernel for it.
